### PR TITLE
refactor!: D6* step() cleanup from pre-post

### PR DESCRIPTION
**Pull request by chenkins**
### Carried out

| ID | Description | Category | Severity | Commit(s) |
|---|---|---|---|---|
| — | Drop the dead `stop_action_given` map-entry candidate for MALFUNCTION_OFF_MAP - `_handle_malfunction_off_map` only ever promotes to MOVING via `movement_action_given`, so the discarded candidate was optimistic dead weight | refactor | Low | f1daa61d |
| — | Dedupe 3 post-hoc position reads against the collect phase's own `candidate_entry_point`, safe now that map-entry candidates are exactly predictive of the real outcome | refactor | Low | 499c13d7 |
| — | Extract `_candidate_entry_points`, reused by `_check_post_position_invariants` - dedupes the check's own position derivation into a shared method; leaves a shadow verification call commented out, having caught a real divergence in the map-entry branch that still needed fixing | refactor | Medium | 147586f6 |
| — | Fix `_candidate_entry_points`' map_entry branch to a state-free, no-+1 formula derived purely from pre-step values + action + an explicit `elapsed_steps` parameter, closing the last remaining violation of the method's pre-step-only input constraint | fix | Medium | da3e1b74 |
| — | Restore `pre_done` in `_candidate_entry_points`, sourced from `agent.target_entry_point` rather than `self.dones` (which `end_of_episode_update()` force-sets True for every agent once the episode ends, not just the one that reached its own target) | fix | Medium | 6c265c3d |
| — | Make `_candidate_entry_points` return the real target-entry candidate instead of `None, None` when `remove_agents_at_target=True`, moving the "expect None" assertion into `_check_post_position_update_invariants` where the post-removal expectation actually belongs; re-enables the collect-phase-vs-`_candidate_entry_points` verification assertion | refactor | Medium | abc92e7c |
| — | Extract `_candidate_distance`/`_candidate_speed` from the post-step speed/distance check's "candidates accepted" branches, replacing assertions with return statements (same pattern as `_candidate_entry_points`); notes both methods still read live/committed state at their call sites, not yet pre-step-only | refactor | Medium | dee641a3 |
| — | Make `_candidate_distance`/`_candidate_speed` derive from pre-step values only, adding a shared `_action_valid` helper computed from `SpeedCounter.is_cell_exit()`'s real state-independent contract instead of reading committed state at the call site | refactor | Medium | 9a38d722 |
| — | Derive `_candidate_distance`/`_candidate_speed`'s invalid-action check from `candidate_entry_point_independent` directly, dropping the `_action_valid` helper's `is_cell_exit` conflation; compute `candidate_entry_point_independent` once per agent per step in `_capture_pre_step_snapshot` instead of once per caller | refactor | Low | 3d3b4ba1 |
| — | Drop step()'s duplicated inline (3a)/(3b) branches, delegating to `_candidate_speed`/`_candidate_entry_points` directly and removing the now-unnecessary verification assert; requires `_candidate_speed` to never return None (done/target-reached/malfunction/stay-off-map branches now uniformly return `Fraction(0)`) | refactor | Medium | bf7287ac |
| — | Replace step()'s (10a) with `_check_post_position_update_invariants`'s own candidate-acceptance structure, installed as assignments instead of assertions; `handle_done_state()` now takes `candidate_entry_point` explicitly instead of reading it back post-nulling | refactor | Medium | ec51dff2 |
| — | Document `is_cell_entry` semantics and add an `xfail(strict=True)` test reproducing a banked-restart boundary gap in `SpeedCounter.step()`'s formula, ahead of replacing it with `set()` | test | Low | afec03fc |
| — | Drive (10b) SPEED_COUNTER UPDATE from the collect phase's own `candidate_speed`/`candidate_distance` (adding (3c) CANDIDATE DISTANCE), replacing `SpeedCounter.step()` with `set(speed, distance)`; fixes the banked-restart `is_cell_entry` gap documented by the previous commit's xfail test | refactor | High | d92e9d86 |
| — | Split (10b)'s off-map-denied case out as its own no-op branch - `speed_counter` is already `(None, None)` there, so the previous ternary was computing an already-in-place value | refactor | Low | 8799efe7 |
| — | Drop the unused `AgentTransitionData.action_valid` field - never read anywhere; every check goes through `state_transition_signal.action_valid` instead | refactor | Low | e11af124 |
| — | Inline `candidate_entry_point`/`candidate_next_entry_point` in `_candidate_entry_points` instead of precomputing them as separate locals; guard target-reached/on-map-transition checks with `is_on_map` explicitly so the arithmetic conjunct is never evaluated with a None pre_speed/pre_offset; flags a known docstring/code mismatch in `_candidate_distance`'s invalid-action branch with a TODO | refactor | Low | 41c06ec9 |
| — | Add a regression test proving a DONE, removed agent's initial-edge lookup is stale (the resource is genuinely free the very next step, no penalty for a contending agent), dropping the now-disproven "if done, we should not try to reserve the initial edge" TODO | test | Medium | e956d1df |
| — | Gate `_candidate_distance`'s invalid-action branch on `is_cell_exit`, mirroring `_candidate_speed`'s own equivalent gate and matching the method's docstring - clarity/consistency fix, not a behavior change (the formulas already coincide off the cell boundary); resolves the "code is unconditional in invalid action" TODO left by 41c06ec9 | fix | Low | dbdd48dd |

### Remaining

No outstanding follow-up work identified for this PR.

## Related issues

Link to every issue from the issue tracker (if any) you addressed in this PR.

## Checklist

- [ ] Tests are included for relevant behavior changes.
- [ ] Documentation is added in the [flatland-book](https://github.com/flatland-association/flatland-book) repo for relevant behavior changes.
- [ ] If you made important user-facing changes, describe them under the `[Unreleased]` tag in `CHANGELOG.md`.
- [ ] New package dependencies are declared in the `pyproject.toml` file.
  Requirement files have been updated by running `tox -e requirements`.
- [ ] Code works with all supported Python versions (3.10, 3.11, 3.12 and 3.13). Checks run with all supported version and are
  required to run successfully.
- [ ] Code is formatted according to PEP 8 (an IDE like PyCharm can do this for you).
- [ ] Technical guidelines listed in `CONTRIBUTING.md` are followed.

### CODE REVIEW

File: flatland/envs/step_utils/speed_counter.py [hunk 1]
```diff
@@ -113,34 +100,41 @@ def __init__(self, max_speed: float, speed: Optional[float] = None):
...
+        # relative to its previous value (a wrap into a new cell), or the agent just bootstrapped onto
+        # the map (previous distance None, new distance concrete). distance never decreases within the
+        # same cell (both distance_after_crossing's modulo and distance_without_crossing's cap only ever
+        # hold or grow it), so "new < old" is unambiguously a crossing, not mid-cell noise. Unlike the
+        # old step()'s new_distance < speed (which reduces to old_distance < SEGMENT_LENGTH under
+        # crossing_completed, wrongly False for an agent banked exactly at the boundary that resumes and
+        # genuinely crosses), this needs no separate crossing_completed flag at all.
+        self._is_cell_entry = (
+            (self._distance is None and distance is not None)
+            or (self._distance is not None and distance is not None and distance < self._distance)
```
Comment (CleverManu) [inline flatland/envs/step_utils/speed_counter.py:129]: What about speed = 1? in this case distance = self._distance
But replacing < by <= doesn't work, since stopped agents satisfy this as well

File: flatland/envs/rail_env.py [hunk 2]
```diff
@@ -794,6 +736,58 @@ def _check_malfunction_state_invariant(self):
     def _is_speed_zero(self, candidate_speed: Fraction) -> bool:
         return candidate_speed == 0.0
 
+    def _candidate_entry_points(self, action: RailEnvActions, agent: EnvAgent,
```
Comment (CleverManu) [inline flatland/envs/rail_env.py:739]: Claude wants to introduce pre_malfunction_off_map: bool) -> Tuple[Optional[EntryPointT], Optional[EntryPointT]] here.

Reason: 
ready_to_depart is explicitly designed (per its own comment) to reconstruct "is the state already READY_TO_DEPART, without reading state" — which is correct only for the WAITING→READY_TO_DEPART path, because that transition has a one-step visibility lag (the state read at the top of a call reflects the previous call's outcome). But this same ready_to_depart value is also used to gate MALFUNCTION_OFF_MAP's map-entry eligibility, and _handle_malfunction_off_map (state_machine.py) checks earliest_departure_reached directly as a same-step signal (elapsed_steps + 1 >= earliest_departure — the state-transition-signal formula, not a lagged reconstruction), with no intervening READY_TO_DEPART step. These are two different questions being answered by one formula. The result: whenever an agent's malfunction ends on exactly the step where earliest_departure == elapsed_steps + 1, the real state machine (computed independently and correctly via agent.state_machine.step()) promotes the agent to MOVING, while _candidate_entry_points — which now drives the actual position update, not just a check, since commit bf7287ac wired it into step()'s real logic — computes (None, None) (stay off-map). The result is agent.state == MOVING with agent.current_entry_point is None, and state_position_sync_check raises ValueError unconditionally (this check is not gated by check_step_pre_post_conditions), crashing env.step().

Comment (CleverManu) [reply flatland/envs/rail_env.py:739]: Claude confirmed this three ways:

- A 500-seed × 8-agent × 80-step fuzz run with malfunctions enabled crashed on 4 seeds (17, 38, 105, 144) with exactly this error, on this branch; the identical script ran clean on all 4 seeds against the PR's own base branch (178-agents-living-on-the-edge-14).
- Bisecting the 18 commits pinpointed the divergence to da3e1b74, with abc92e7c producing a self-diagnosing assertion: _candidate_entry_points diverged from inline (3b): inline=(((8, 11), 0), ((7, 11), 0)) _candidate_entry_points=(None, None) agent=0 state=2 in_malfunction=False action=MOVE_RIGHT.
- A minimal, deterministic single-agent repro (no randomness):
```
agent.earliest_departure = 6
agent.malfunction_handler._set_malfunction_down_counter(5)  # malfunction ends exactly step 5
for i in range(1, 6):
    env.step({agent.handle: RailEnvActions.MOVE_FORWARD})
# ValueError on step 5: "Agent State 3 is on map Agent Position None if off map"
```

Comment (CleverManu) [reply flatland/envs/rail_env.py:739]: Sorry for the long comment, but a suggested change would be quite far reaching, so I wanted to bring up the issue this way

File: tests/test_flatland_envs_rail_env.py [hunk 3]
```diff
@@ -736,6 +737,216 @@ def test_symmetric_switch_move_forward_action():
...
+    L = ((3, 8), Grid4TransitionsEnum.WEST)
+    agent.current_entry_point = L
+    agent._set_state(TrainState.MOVING)
+    agent.next_entry_point = _sanitize_entry_point(env.rail.apply_action_independent(RailEnvActions.MOVE_FORWARD, L))
+    agent.speed_counter = SpeedCounter(max_speed=Fraction(1), speed=Fraction(1))
+    agent.speed_counter.set(speed=Fraction(1), distance=Fraction(1, 2))
+
+    resource_check = _assert_speed_distance_match_candidates(env, agent, {0: RailEnvActions.MOVE_FORWARD})
+    assert resource_check
+    assert agent.current_entry_point != L  # the crossing genuinely completed
```
Comment (CleverManu) [inline tests/test_flatland_envs_rail_env.py:818]:
```suggestion
    assert agent.current_entry_point != L  # crossing completed
```

File: tests/test_flatland_envs_rail_env.py [hunk 4]
```diff
@@ -736,6 +737,216 @@ def test_symmetric_switch_move_forward_action():
...
+    can be blocked by.
+
+    Rail: `_make_straight_rail(3)`'s 3-cell corridor A=(0,0) - B=(0,1) - C=(0,2). Agent 0's target is B;
+    B is also agent 1's own initial cell (agent 1's line runs B -> C).
+
+    - Setup: agent 0 bootstrapped directly onto A, MOVING at max_speed=1 with distance already at the
+      cell boundary (crosses into B in one step). Agent 1 stays off map, READY_TO_DEPART with
+      earliest_departure=0 and initial_entry_point B.
+    - Step 0: both agents contest resource B in the same step - agent 0 crossing A->B, agent 1 departing
+      directly onto B - so agent 0 genuinely occupies B (agent 1's own initial cell) as this step is
```
Comment (CleverManu) [inline tests/test_flatland_envs_rail_env.py:896]:
```suggestion
      directly onto B - so agent 0 occupies B (agent 1's own initial cell) as this step is
```

**Review by CleverManu — APPROVED** [review]

File: flatland/envs/step_utils/speed_counter.py (hunk 1 above)
Comment (chenkins) [reply flatland/envs/step_utils/speed_counter.py:129]: Fixed (d92e9d86 - is_cell_entry rewritten to compare distance < self._distance instead of new_distance < speed, so it no longer depends on speed at all - genuinely resolves the speed=1 edge case)

Comment (chenkins) [reply flatland/envs/step_utils/speed_counter.py:129]: @CleverManu please verify whether this is true.

File: tests/test_flatland_envs_rail_env.py (hunk 4 above)
Comment (chenkins) [reply tests/test_flatland_envs_rail_env.py:896]: Fixed 052f82d39b1ca5c3a2fd19e0a2530fca01e530e1

File: tests/test_flatland_envs_rail_env.py (hunk 3 above)
Comment (chenkins) [reply tests/test_flatland_envs_rail_env.py:818]: Fixed 052f82d39b1ca5c3a2fd19e0a2530fca01e530e1

File: flatland/envs/rail_env.py (hunk 2 above)
Comment (chenkins) [reply flatland/envs/rail_env.py:739]: Fixed by https://github.com/flatland-association/flatland-rl/pull/524, please double-check.
